### ISSUE TITLE
Allow to programmatically ignore errors (in Grape)

### ISSUE THREAD

**Issue by lvonk** [appsignal/appsignal-ruby#586]
Hi,

In the Sinatra integration (when disabling `raise_errors`) it is possible to programmatically ignore errors that are send to AppSignal by setting `env['sinatra.skip_appsignal_error']`:

https://github.com/appsignal/appsignal-ruby/blob/c9fa71b056320cfe2cdd68aa57869f08e4545cf9/lib/appsignal/rack/sinatra_instrumentation.rb#L70

This is great since it allows a more fine-grained control (than specifying them in `ignore_errors` in the `appsignal.yml`) over what is send to AppSignal and what not.

This feature is however not supported in Grape, in fact it seems it is only supported in Sinatra. Is this something that can be added for Grape (and possibly all integrations?)

**Comment by tombruijn** [comment]
Responded in https://app.intercom.io/a/apps/yzor8gyw/inbox/inbox/unassigned/conversations/25888280844

**Comment by lvonk** [comment]
Hi,

So when trying to add this to the appsignal gem it turned out to be a bit more complex than I initially thought. Due to the order of middlewares: `insert_before Grape::Middleware::Error, Appsignal::Grape::Middleware` the Appsignal wraps the  `Grape::Middleware::Error`. If you then implement `rescue_from` and return a response to the user there is no error anymore for the appsignal middleware to handle. By implementing rescue_from as below you have the same result as Sinatra:

```ruby
base.rescue_from :all do |e|
    env['grape.error'] = e
end
```

And then in the appsignal middleware:

```ruby
if env['grape.error'] && !env['grape.skip_appsignal_error']
  transaction.set_error(env['grape.error'])
end
```

**Comment by tombruijn** [comment]
Hi @lvonk,

I've created PR #588 that should implement what you suggested. Can you confirm this works for you?

I'm not sure what the scenario you described in your comment. If we use `insert_before` we should capture the error before Grape renders an error page. The AppSignal middleware doesn't wrap the Grape middleware, but the other way around.

**Comment by lvonk** [comment]
Hmm strange, when I rescue the error in grape like:

```ruby
base.rescue_from :all do |e|
  error!({foo: 'bar'}.to_json, 500)
end
```

with
```ruby
insert_before Grape::Middleware::Error, Appsignal::Grape::Middleware
```

Then it will reach the Appsignal Middleware rescue block (since the error is handled):

https://github.com/appsignal/appsignal-ruby/blob/7ab700d6dd37720b9062fa6387dc5166b9afc6e8/lib/appsignal/integrations/grape.rb#L23-L28

**Comment by tombruijn** [comment]
Hi @lvonk,

> Is this also what you are seeing?

Yes, if I add my own `base.rescue_from :all` block I do not see it reach the AppSignal Grape middleware, probably because the error has been rescue-ed and does not bubble up the stack any further.

> Since we could also add `Appsignal.add_exception(e)` in those rescue blocks in which we do want the exception to be logged instead of stating in which rescue blocks we do not want it to be logged.

That would be the way to go, manually set the error on the AppSignal transaction if you the error catch in a `base.rescue_from` block.

**Comment by tombruijn** [comment]
The suggested fix has been shipped in Ruby gem 2.10.5. Docs can be found here: https://docs.appsignal.com/ruby/integrations/grape.html#ignoring-errors
